secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,229 @@
+        missingIndices: [Int],
+        missingWordsChips: [PhraseChip.Kind]
+    ) -> Self {
+        RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordsChips,
+            completion: []
+        )
+    }
+
+    /// creates an initial `RecoveryPhraseValidationState` with no completions and random missing indices.
+    /// - Note: Use this function to create a random validation puzzle for a given phrase.
+    static func random(phrase: RecoveryPhrase) -> Self {
+        let missingIndices = Self.randomIndices()
+        let missingWordChipKind = Self.pickWordsFromMissingIndices(indices: missingIndices, phrase: phrase).shuffled()
+        return RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordChipKind,
+            completion: []
+        )
+    }
+}
+
+extension RecoveryPhraseValidationState {
+    /// Given an array of RecoveryPhraseStepCompletion, missing indices, original phrase and the number of groups it was split into,
+    /// assembly the resulting phrase. This comes up with the "proposed solution" for the recovery phrase validation challenge.
+    /// - returns:an array of String containing the recovery phrase words ordered by the original phrase order.
+    static func resultingPhrase(
+        from completion: [RecoveryPhraseStepCompletion],
+        missingIndices: [Int],
+        originalPhrase: RecoveryPhrase,
+        numberOfGroups: Int
+    ) -> [String] {
+        precondition(missingIndices.count == completion.count)
+        precondition(completion.count == numberOfGroups)
+
+        var words = originalPhrase.words
+        let groupLength = words.count / numberOfGroups
+        // iterate based on the completions the user did on the UI
+        for wordCompletion in completion {
+            // figure out which phrase group (chunk) this completion belongs to
+            let i = wordCompletion.groupIndex
+            // validate that's the right number
+            precondition(i < numberOfGroups)
+            // get the missing index that the user did this completion for on the given group
+            let missingIndex = missingIndices[i]
+            // figure out what this means in terms of the whole recovery phrase
+            let concreteIndex = i * groupLength + missingIndex
+            precondition(concreteIndex < words.count)
+            // replace the word on the copy of the original phrase with the completion the user did
+            words[concreteIndex] = wordCompletion.word
+        }
+
+        return words
+    }
+
+    /// drives the state machine represented on this Enum by the action of applying a chip into a group of words containing an empty slot that has to be completed
+    static func given(_ state: RecoveryPhraseValidationState, apply chip: PhraseChip.Kind, into group: Int) -> Self {
```

Comment:
Instead of a static function on State that has to take in the State, this could just be a regular function on State I believe.

Reply:
renamed to apply(chip:into:)